core/logger/zap.go

```diff
@@ -1,31 +1,129 @@
 package logger
 
 import (
-	"errors"
 	"fmt"
 	"io"
 	"os"
+	"sync"
 
+	"github.com/smartcontractkit/chainlink/core/utils"
+
+	"github.com/pkg/errors"
 	"go.uber.org/zap"
 	"go.uber.org/zap/zapcore"
 )
 
 var _ Logger = &zapLogger{}
 
+// ZapLoggerConfig defines the struct that serves as config when spinning up a the zap logger
+type ZapLoggerConfig struct {
+	zap.Config
+	local          Config
+	diskLogLevel   zap.AtomicLevel
+	sinks          []zapcore.WriteSyncer
+	diskStats      utils.DiskStatsProvider
+	diskPollConfig zapDiskPollConfig
+	diskLogLvlChan chan zapcore.Level
+}
+
 type zapLogger struct {
 	*zap.SugaredLogger
-	config     zap.Config
-	name       string
-	fields     []interface{}
-	callerSkip int
+	config            ZapLoggerConfig
+	name              string
+	fields            []interface{}
+	callerSkip        int
+	closeDiskPollChan chan struct{}
 }
 
-func newZapLogger(cfg zap.Config) (Logger, error) {
-	zl, err := cfg.Build()
-	if err != nil {
-		return nil, err
+func newZapLogger(cfg ZapLoggerConfig) (Logger, func() error, error) {
+	cfg.diskLogLevel = zap.NewAtomicLevelAt(zapcore.DebugLevel)
+
+	cores := []zapcore.Core{
+		newConsoleCore(cfg),
+	}
+	newCores, err := newCores(cfg)
+	cores = append(cores, newCores...)
+
+	core := zapcore.NewTee(cores...)
+	lggr := &zapLogger{
+		config:            cfg,
+		closeDiskPollChan: make(chan struct{}),
+		SugaredLogger:     zap.New(core).Sugar(),
+	}
+
+	if cfg.local.ToDisk {
+		go lggr.pollDiskSpace()
+	}
+
+	var once sync.Once
+	close := func() error {
+		once.Do(func() {
+			if cfg.local.ToDisk {
+				close(lggr.closeDiskPollChan)
```

Comment:
Can we make this a synchronous close pls e.g. with a ping-pong channel or waitgroup?

Not a big deal but you can end up with flakey tests and other annoyances if e.g. the application has exited but it left a goroutine hanging around that didn't finish up yet.